Mathlib/Analysis/NormedSpace/Connected.lean

```diff
@@ -0,0 +1,202 @@
+  let z := c + t • y
+  simp only [compl_union, mem_inter_iff, mem_compl_iff, mem_setOf_eq, not_nonempty_iff_eq_empty]
+    at ht
+  have JA : JoinedIn sᶜ a z := by
+    apply JoinedIn_of_segment_subset
+    rw [subset_compl_iff_disjoint_right, disjoint_iff_inter_eq_empty]
+    convert ht.2
+    exact Ia.symm
+  have JB : JoinedIn sᶜ b z := by
+    apply JoinedIn_of_segment_subset
+    rw [subset_compl_iff_disjoint_right, disjoint_iff_inter_eq_empty]
+    convert ht.1
+    exact Ib.symm
+  exact JA.trans JB.symm
+
+/-- In a real vector space of dimension `> 1`, the complement of any countable set is
+connected. -/
+theorem Set.Countable.isConnected_compl_of_one_lt_rank (h : 1 < Module.rank ℝ E) {s : Set E}
+    (hs : s.Countable) : IsConnected sᶜ :=
+  (hs.isPathConnected_compl_of_one_lt_rank h).isConnected
+
+/-- In a real vector space of dimension `> 1`, the complement of a singleton is path
+connected. -/
+theorem isPathConnected_compl_singleton_of_one_lt_rank (h : 1 < Module.rank ℝ E) (x : E) :
+    IsPathConnected {x}ᶜ :=
+  Set.Countable.isPathConnected_compl_of_one_lt_rank h (countable_singleton x)
+
+/-- In a real vector space of dimension `> 1`, the complement of a singleton is
+connected. -/
+theorem isConnected_compl_singleton_of_one_lt_rank (h : 1 < Module.rank ℝ E) (x : E) :
+    IsConnected {x}ᶜ :=
+  (isPathConnected_compl_singleton_of_one_lt_rank h x).isConnected
+
+end TopologicalVectorSpace
+
+section NormedSpace
+
+variable {E : Type*} [NormedAddCommGroup E] [NormedSpace ℝ E]
+
+/-- In a real vector space of dimension `> 1`, any sphere of nonnegative radius is
+path connected. -/
+theorem isPathConnected_sphere (h : 1 < Module.rank ℝ E) (x : E) {r : ℝ} (hr : 0 ≤ r) :
+    IsPathConnected (sphere x r) := by
+  /- when `r > 0`, we write the sphere as the image of `{0}ᶜ` under the map
+  `y ↦ x + (r * ‖y‖⁻¹) • y`. Since the image under a continuous map of a path connected set
+  is path connected, this concludes the proof. -/
+  rcases hr.eq_or_lt with rfl|rpos
+  · simpa using isPathConnected_singleton x
+  let f : E → E := fun y ↦ x + (r * ‖y‖⁻¹) • y
+  have A : ContinuousOn f {0}ᶜ := by
+    intro y hy
+    apply (continuousAt_const.add _).continuousWithinAt
+    apply (continuousAt_const.mul (ContinuousAt.inv₀ continuousAt_id.norm ?_)).smul continuousAt_id
+    simpa using hy
+  have B : IsPathConnected ({0}ᶜ : Set E) := isPathConnected_compl_singleton_of_one_lt_rank h 0
+  have C : IsPathConnected (f '' {0}ᶜ) := B.image' A
+  have : f '' {0}ᶜ = sphere x r := by
+    apply Subset.antisymm
+    · rintro - ⟨y, hy, rfl⟩
+      have : ‖y‖ ≠ 0 := by simpa using hy
+      simp [norm_smul, abs_of_nonneg hr, mul_assoc, inv_mul_cancel this]
+    · intro y hy
+      refine ⟨y - x, ?_, ?_⟩
+      · intro H
+        simp only [mem_singleton_iff, sub_eq_zero] at H
+        simp only [H, mem_sphere_iff_norm, sub_self, norm_zero] at hy
+        exact rpos.ne hy
+      · simp at hy
```

Comment:
We should really squeeze this:
```suggestion
      · simp at hy
```
(or something else like `replace hy : ... := by simpa using hy`)